File: #134_Gas_Station/Solution.java
public class Solution {
	public int canCompleteCircuit(int[] gas, int[] cost) {
		if (gas == null || gas.length == 0 || cost == null || cost.length == 0) return -1;
		int len = gas.length;
		// start from gas[0]
		int index = 0;
		while (index < len) {
			int cur = index;
			int r = 0;
			int step = 0;
			while (r + gas[cur] >= cost[cur]) {
				step++;
				if (step == len) return index;
				r = r + gas[cur] - cost[cur];
				cur++;
				if (cur == len) cur = 0;
			}
			if (index + step >= len) break;
			index += step + 1;
		}
		return -1;
	}

	public static void main(String[] args) {
		Solution sol = new Solution();
		int[] g = {4};
		int[] c = {5};
		System.out.println(sol.canCompleteCircuit(g, c));
	}
}

File: #134_Gas_Station/Solution.py
class Solution:
	# @param {integer[]} gas
	# @param {integer[]} cost
	# @return {integer}
	def canCompleteCircuit(self, gas, cost):
		if not len(gas) or not len(cost):
			return -1
		if len(gas) != len(cost):
			return -1
		index = 0
		while index < len(gas):  # don't mod index ever
			r = 0
			cur = index
			step = 0
			while r + gas[cur] >= cost[cur]:
				step += 1
				r = r + gas[cur] - cost[cur]
				cur = (cur + 1) % len(gas)
				if step == len(gas):
					return index
			if index + step >= len(gas):
				return -1
			index += step + 1    # new index
		return -1

gas = [5]
cost = [4]
sol = Solution()
print sol.canCompleteCircuit(gas, cost)